icu4c/source/i18n/smpdtfmt.cpp

```diff
@@ -2575,51 +2582,47 @@ SimpleDateFormat::parse(const UnicodeString& text, Calendar& cal, ParsePosition&
             } else { // tztype == TZTYPE_DST
                 if (dst == 0) {
                     if (btz != NULL) {
```

Comment:
TZ database 2020e added daylight saving rules with 30 minutes offset around 1947 to 1968 in America/Belize.  This change revealed a design issue of "golden zone", which is used for mapping metazone (e.g. America_Central) back to timezone (e.g. America/Chicago for US, America/Belize for Belize).

America/Chicago is global golden zone for metazone America_Central. America/Belize is regional golden zone for metazone America_Central. When a date string "1970-07-01 12:00:00 Central Daylight Time" is parsed by ICU date format initialized with en_US locale, "Central Daylight Time" is mapped to metazone America_Central, then mapped to time zone America/Chicago. Chicago observed daylight saving time on July 1st, 1970, there are no problems parsing the strings into date value. However, if ICU date format is initialized with en_BZ locale, the metazone is mapped to time zone America/Belize. Belize did not observe daylight saving time on July 1st, 1970, so this implementation tries to adjust offset (by parsing "Central Daylight Time", DST is detected, but time zone offset of America/Belize has no daylight offset on the day).

Previous logic for adjusting daylight offset was based on nearest daylight saving time before and after the given time. tzdata up to 2020d did not have daylight saving rule up to 1968 with +30 minutes offset. Belize used +1 hour daylight offset in 1973, so this value was used with 2020d. As the result, it's still consistent with all other time zones mapped to America_Central.

tzdata 2020e added the irregular daylight saving time (+30 minutes offset) in 1968, so the old implementation evaluates daylight offset in Feb 1968 (+30 minutes) and Dec 1973 (+ 1 hour). In this case, Feb 1968 is closer, and the offset at that time is picked.

The new implementation changed precedence. When daylight saving time is not observed by the resolved time zone, it looks for future daylight saving time. It only looks for past daylight saving time only when daylight saving time was never used after the parsed time. This implementation works better for a few reasons.

- CLDR metazone based display names are never used for date before 1970.
- If Belize uses odd daylight saving offset after 1970, ICU does not use the metazone America_Central for the period, so this is not an issue.
- In general, there are fewer zones using daylight offset other than +1 hour in modern dates.

Strictly speaking, this is a limitation of the current golden zone concept. If we need to support parsing date string with zone name across regions, we need some extra data/mechanism. I'm not really proposing to improve the design, because we need to pay extra overhead (more data, more complicated code) just for resolving some edge case issues.